ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,647 @@
+
+      int label;
+      try
+      {
+        label = std::get<int>(labelAny);
+      }
+      catch(std::bad_variant_access &e)
+      {
+        // items with no class are considered background
+        label = this->backgroundLabel;
+      }
+
+      for (unsigned int i = 0; i < item->getNumSubItems(); i++)
+      {
+        // save subitems material
+        Ogre::SubItem *subItem = item->getSubItem(i);
+        Ogre::HlmsDatablock *datablock = subItem->getDatablock();
+        this->datablockMap[subItem] = datablock;
+
+        // Material Switching
+        if (this->type == SegmentationType::Semantic)
+        {
+          if (this->isColoredMap)
+          {
+            // semantic material(each pixel has item's color)
+            math::Color color = this->LabelToColor(label);
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              color.R(), color.G(), color.B(), 1.0));
+          }
+          else
+          {
+            // labels ids material(each pixel has item's label)
+            float labelColor = label / 255.0;
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              labelColor, labelColor, labelColor, 1.0));
+          }
+        }
+        else if (this->type == SegmentationType::Panoptic)
+        {
+          if (!this->instancesCount.count(label))
+            this->instancesCount[label] = 0;
+
+          this->instancesCount[label]++;
+          int instanceCount = this->instancesCount[label];
+
+          if (this->isColoredMap)
+          {
+            // convert 24 bit number to int64
+            int compositeId = label * 256 * 256 + instanceCount;
+
+            math::Color color;
+            if (label == this->backgroundLabel)
+              color = this->LabelToColor(label);
+            else
+              color = this->LabelToColor(compositeId);
+
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              color.R(), color.G(), color.B(), 1.0));
+          }
+          else
+          {
+            // 256 => 8 bits .. 255 => color percentage
+            float labelColor = label / 255.0;
+            float instanceColor1 = (instanceCount / 256) / 255.0;
+            float instanceColor2 = (instanceCount % 256) / 255.0;
+
+            subItem->setCustomParameter(1, Ogre::Vector4(
+              labelColor, instanceColor1, instanceColor2, 1.0));
+          }
+        }
+
+        // check if it's an overlay material by assuming the
+        // depth check and depth write properties are off.
+        if (!datablock->getMacroblock()->mDepthWrite &&
+            !datablock->getMacroblock()->mDepthCheck)
+          subItem->setMaterial(this->plainOverlayMaterial);
+        else
+          subItem->setMaterial(this->plainMaterial);
+      }
+    }
+    itor.moveNext();
+  }
+
+  // reset the count & colors tracking
+  this->instancesCount.clear();
+  this->takenColors.clear();
+  this->coloredLabel.clear();
```

Comment:
Is there a reason why we have to reset this after every pre-render? This makes me think that we'd have to re-compute colors each time around, and since we are using a random number generator to compute colors, this could mean that we don't have consistent colors for objects on each update.

Reply:
Yeah, It is calculated each iteration. but the same colors will be produced for the objects !
because in the random generator, I use the unique encoded ID of each item as a seed. so it generates the same color each time